src/vs/platform/actionWidget/browser/actionWidget.ts

```diff
@@ -0,0 +1,619 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+import { ActionBar } from 'vs/base/browser/ui/actionbar/actionbar';
+import * as dom from 'vs/base/browser/dom';
+import { IAnchor } from 'vs/base/browser/ui/contextview/contextview';
+import { KeybindingLabel } from 'vs/base/browser/ui/keybindingLabel/keybindingLabel';
+import { IListEvent, IListMouseEvent, IListRenderer, IListVirtualDelegate } from 'vs/base/browser/ui/list/list';
+import { List } from 'vs/base/browser/ui/list/listWidget';
+import { IAction } from 'vs/base/common/actions';
+import { Disposable, DisposableStore, IDisposable, MutableDisposable } from 'vs/base/common/lifecycle';
+import { OS } from 'vs/base/common/platform';
+import { localize } from 'vs/nls';
+import { ICommandService } from 'vs/platform/commands/common/commands';
+import { IContextKeyService, RawContextKey } from 'vs/platform/contextkey/common/contextkey';
+import { IContextViewService } from 'vs/platform/contextview/browser/contextView';
+import { IKeybindingService } from 'vs/platform/keybinding/common/keybinding';
+import { ITelemetryService } from 'vs/platform/telemetry/common/telemetry';
+import { Codicon } from 'vs/base/common/codicons';
+import 'vs/css!./actionWidget';
+import { createDecorator, IInstantiationService, ServicesAccessor } from 'vs/platform/instantiation/common/instantiation';
+import { InstantiationType, registerSingleton } from 'vs/platform/instantiation/common/extensions';
+import { Action2, registerAction2 } from 'vs/platform/actions/common/actions';
+import { KeyCode, KeyMod } from 'vs/base/common/keyCodes';
+import { KeybindingWeight } from 'vs/platform/keybinding/common/keybindingsRegistry';
+import { ActionSet, IActionItem, IActionKeybindingResolver } from 'vs/platform/actionWidget/common/actionWidget';
+
+export const acceptSelectedActionCommand = 'acceptSelectedCodeAction';
+export const previewSelectedActionCommand = 'previewSelectedCodeAction';
+
+export const ActionWidgetContextKeys = {
+	Visible: new RawContextKey<boolean>('actionWidgetVisible', false, localize('actionWidgetVisible', "Whether the action widget list is visible"))
+};
+export interface IRenderDelegate<T extends IActionItem> {
+	onHide(didCancel?: boolean): void;
+	onSelect(action: IActionItem, preview?: boolean): Promise<any>;
+}
+
+export interface IActionShowOptions {
+	readonly includeDisabledActions: boolean;
+	readonly fromLightbulb?: boolean;
+	readonly showHeaders?: boolean;
+}
+
+export interface IListMenuItem<T extends IActionItem> {
+	item?: T;
+	kind?: any;
```

Comment:
Do we need `kind` here anymore?

Reply:
yes we do because this refers to the kind of menu item rather than the kind of group so determines which renderer gets used

Reply:
If it's just determining header vs action this could be a boolean or use a const enum with 2 values.

Reply:
that complicates things imo

Reply:
it is already a const enum

```
export const enum ActionListItemKind {
	Action = 'action',
	Header = 'header'
}

```

Reply:
i'll just set that as the type for kind